File: tests/AlertAudioTests.cs
using System;
using System.IO;
using FreeIsland;

internal static class AlertAudioTests
{
    private static int checks;
    private static void Check(bool value, string message) { if (!value) throw new Exception(message); ++checks; }
    private static void Reject(Action action, string message) { bool rejected = false; try { action(); } catch (ArgumentException) { rejected = true; } Check(rejected, message); }
    private sealed class Output : IAlertAudioOutput
    {
        public int Starts, Stops, Disposals; public bool Fail; public double Volume; public string Path; public Action Failed, Ended;
        public void Play(string path, double volume, Action failed, Action ended) { ++Starts; Path = path; Volume = volume; Failed = failed; Ended = ended; if (Fail) throw new InvalidOperationException("fixture decoder failure"); }
        public void Stop() { ++Stops; }
        public void SetVolume(double volume) { Volume = volume; }
        public void Dispose() { ++Disposals; }
    }
    [STAThread] private static int Main(string[] args)
    {
        try {
            if (args.Length != 1) throw new Exception("Provide isolated workspace fixture directory");
            string root = Path.GetFullPath(args[0]); Directory.CreateDirectory(root);
            string wav = Path.Combine(root, "课堂铃声.wav"), mp3 = Path.Combine(root, "提醒音频.MP3"), other = Path.Combine(root, "not-audio.exe");
            File.WriteAllText(wav, "fixture, never sent to a real decoder"); File.WriteAllText(mp3, "fixture, never sent to a real decoder"); File.WriteAllText(other, "fixture");
            var output = new Output(); int fallback = 0; DateTime now = new DateTime(2026, 1, 1);
            string settings = Path.Combine(root, Guid.NewGuid().ToString("N"));
            using (var audio = new AlertAudioService(settings, false, output, delegate { ++fallback; }, delegate { return now; }, false)) {
                Check(audio.VolumePercent == 70 && audio.GetPath(AlertSoundKind.Countdown) == "", "Default audio preferences");
                audio.Play(AlertSoundKind.Countdown); Check(fallback == 1 && output.Starts == 0 && !audio.IsPlaying, "Default sound once");
                Reject(delegate { audio.SetPath(AlertSoundKind.Countdown, "https://example.test/a.wav"); }, "URL rejected");
                Reject(delegate { audio.SetPath(AlertSoundKind.Countdown, "\\\\server\\share\\a.wav"); }, "Network path rejected");
                Reject(delegate { audio.SetPath(AlertSoundKind.Countdown, "\\\\?\\C:\\a.wav"); }, "Device path rejected");
                Reject(delegate { audio.SetPath(AlertSoundKind.Countdown, wav + ":evil.mp3"); }, "Alternate data stream rejected");
                Reject(delegate { audio.SetPath(AlertSoundKind.Countdown, other); }, "Executable rejected");
                Reject(delegate { audio.SetPath(AlertSoundKind.Countdown, Path.Combine(root, "missing.wav")); }, "Missing file rejected");
                Reject(delegate { audio.GetPath((AlertSoundKind)99); }, "Invalid kind rejected");
                audio.SetPath(AlertSoundKind.Countdown, wav); audio.SetPath(AlertSoundKind.Reminder, mp3); audio.SetPath(AlertSoundKind.Shutdown, wav); audio.VolumePercent = 42; audio.Save();
                using (var restored = new AlertAudioService(settings, true, new Output(), delegate { throw new Exception("Unexpected sound"); }, delegate { return now; }, false)) {
                    Check(restored.GetPath(AlertSoundKind.Countdown) == wav && restored.GetPath(AlertSoundKind.Reminder) == mp3 && restored.GetPath(AlertSoundKind.Shutdown) == wav && restored.VolumePercent == 42, "Unicode paths, three events and volume survive restart");
                }
                audio.Play(AlertSoundKind.Countdown); Check(output.Starts == 1 && output.Path == wav && output.Volume == .42 && audio.IsPlaying, "Custom countdown uses selected path and volume");
                Action staleFailure = output.Failed, staleEnd = output.Ended;
                audio.Play(AlertSoundKind.Reminder); Check(output.Starts == 2 && output.Path == mp3, "New event replaces previous audio");
                staleFailure(); staleEnd(); Check(audio.IsPlaying && fallback == 1, "Late callbacks cannot stop replacement or beep");
                audio.StopShutdown(); Check(audio.IsPlaying, "Cancel shutdown preserves unrelated reminder");
                now = now.AddSeconds(29); audio.Tick(); Check(audio.IsPlaying, "Ordinary alert remains before cap");
                now = now.AddSeconds(1); audio.Tick(); Check(!audio.IsPlaying, "Ordinary alert stops after 30 seconds");
                audio.Play(AlertSoundKind.Shutdown); now = now.AddSeconds(10); audio.Tick(); Check(!audio.IsPlaying, "Shutdown warning cap is 10 seconds");
                audio.Play(AlertSoundKind.Shutdown); audio.StopShutdown(); Check(!audio.IsPlaying, "Cancel or fire stops shutdown warning immediately");
                audio.Play(AlertSoundKind.Shutdown, true); audio.StopShutdown(); Check(audio.IsPlaying, "Explicit shutdown audio preview is independent of shutdown plan state"); audio.Stop();
                audio.Play(AlertSoundKind.Reminder, true); now = now.AddSeconds(8); audio.Tick(); Check(!audio.IsPlaying, "Preview capped to 8 seconds");
                audio.Play(AlertSoundKind.Reminder); output.Ended(); Check(!audio.IsPlaying, "Completed media is released without loop");
                audio.Play(AlertSoundKind.Reminder); output.Failed(); int once = fallback; output.Failed(); Check(!audio.IsPlaying && fallback == once && audio.LastError.Length > 0, "Asynchronous decoder error falls back only once");
                output.Fail = true; audio.Play(AlertSoundKind.Reminder); Check(fallback == once + 1 && !audio.IsPlaying, "Synchronous decoder error falls back"); output.Fail = false;
                File.Delete(mp3); audio.Play(AlertSoundKind.Reminder); Check(fallback == once + 2 && !audio.IsPlaying && audio.LastError.Length > 0, "Removed file falls back and keeps selection");
                Check(audio.GetPath(AlertSoundKind.Reminder) == mp3, "Missing removable file choice retained");
                audio.Play(AlertSoundKind.Countdown); audio.VolumePercent = 0; int starts = output.Starts; once = fallback; audio.Play(AlertSoundKind.Shutdown); Check(!audio.IsPlaying && starts == output.Starts && once == fallback, "Zero volume stops and mutes both custom and default sounds");
                audio.VolumePercent = 200; Check(audio.VolumePercent == 100 && output.Volume == 1, "High volume clamped"); audio.VolumePercent = -10; Check(audio.VolumePercent == 0, "Negative volume clamped");
                audio.SetPath(AlertSoundKind.Countdown, ""); Check(audio.GetPath(AlertSoundKind.Countdown) == "", "Reset event to default");
            }
            Check(output.Disposals == 1, "Owned output disposed once");
            var safeOutput = new Output(); int safeFallback = 0;
            using (var safeAudio = new AlertAudioService(Path.Combine(root, "safe"), true, safeOutput, delegate { ++safeFallback; }, delegate { return now; }, false)) {
                safeAudio.Play(AlertSoundKind.Countdown); safeAudio.SetPath(AlertSoundKind.Countdown, wav); safeAudio.Play(AlertSoundKind.Countdown, true);
                Check(safeOutput.Starts == 0 && safeFallback == 0 && !safeAudio.IsPlaying, "Safe mode suppresses default, custom and preview audio");
            }
            File.WriteAllText(Path.Combine(settings, "alert-audio.json"), "broken JSON");
            using (var corrupt = new AlertAudioService(settings, true, new Output(), delegate { }, delegate { return now; }, false)) Check(corrupt.VolumePercent == 70 && corrupt.LastError.Length > 0, "Corrupt settings recover defaults");
            File.WriteAllText(Path.Combine(settings, "alert-audio.json"), "{}");
            using (var partial = new AlertAudioService(settings, true, new Output(), delegate { }, delegate { return now; }, false)) Check(partial.VolumePercent == 70 && partial.GetPath(AlertSoundKind.Countdown) == "", "Missing preference members retain defaults");
            Console.WriteLine("PASS: " + checks + " alert audio checks; injected playback only, no real sound."); return 0;
        } catch (Exception error) { Console.Error.WriteLine("FAIL: " + error); return 1; }
    }
}

File: win7/tests/alert_audio_tests.cpp
#include "../src/alert_audio.h"
#include <iostream>
#include <stdexcept>
#include <memory>

static int checks = 0;
static void Check(bool value, const char* message) { if (!value) throw std::runtime_error(message); ++checks; }
static void Put(const std::wstring& path, const std::string& content) { HANDLE h = CreateFileW(path.c_str(), GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, 0, NULL); Check(h != INVALID_HANDLE_VALUE, "Fixture file opened"); DWORD written = 0; bool ok = WriteFile(h, content.data(), static_cast<DWORD>(content.size()), &written, NULL) && written == content.size(); CloseHandle(h); Check(ok, "Fixture file written"); }
struct Output : fi::AlertAudioOutput {
    int starts = 0, stops = 0, volume = 0; bool fail = false, playing = false; std::wstring path;
    bool Play(const std::wstring& p, int v) override { ++starts; path = p; volume = v; playing = !fail; return !fail; }
    void SetVolume(int v) override { volume = v; }
    void Stop() override { ++stops; playing = false; }
    bool Playing() override { return playing; }
};
int wmain(int argc, wchar_t** argv) {
    try {
        Check(argc == 2, "Provide fresh isolated fixture directory"); std::wstring root = argv[1]; Check(CreateDirectoryW(root.c_str(), NULL) != FALSE, "Unique fixture root");
        std::wstring wav = root + L"\\课堂铃声.wav", mp3 = root + L"\\提醒音频.MP3"; Put(wav, "fixture, no real decoder"); Put(mp3, "fixture, no real decoder");
        using Kind = fi::AlertSoundKind; int fallback = 0; ULONGLONG now = 100000; Output* out = new Output();
        fi::AlertAudio audio(root, false, std::unique_ptr<fi::AlertAudioOutput>(out), [&] { ++fallback; }, [&] { return now; });
        Check(audio.VolumePercent() == 70 && audio.Path(Kind::Countdown).empty(), "Defaults"); audio.Play(Kind::Countdown); Check(fallback == 1 && !out->starts, "Default sound once");
        for (const wchar_t* p : {L"https://example.test/a.wav", L"\\\\server\\share\\a.wav", L"\\\\?\\C:\\a.wav", L"C:\\a.wav:evil.mp3", L"C:\\a.exe", L"C:\\x\r\nvolume=100.wav"}) Check(!audio.SetPath(Kind::Countdown, p), "Unsupported or injected path rejected");
        Check(!audio.SetPath(Kind::Countdown, root + L"\\missing.wav"), "Missing file rejected");
        Check(audio.SetPath(Kind::Countdown, wav) && audio.SetPath(Kind::Reminder, mp3) && audio.SetPath(Kind::Shutdown, wav), "Three event choices"); audio.SetVolumePercent(42); Check(audio.Save(), "Atomic configuration save");
        { fi::AlertAudio restored(root, true); Check(restored.Path(Kind::Countdown) == wav && restored.Path(Kind::Reminder) == mp3 && restored.Path(Kind::Shutdown) == wav && restored.VolumePercent() == 42, "Unicode paths and volume restored"); }
        audio.Play(Kind::Countdown); Check(out->starts == 1 && out->path == wav && out->volume == 42 && audio.IsPlaying(), "Custom countdown playback");
        audio.Play(Kind::Reminder); Check(out->starts == 2 && out->path == mp3, "New event replaces previous audio"); audio.StopShutdown(); Check(audio.IsPlaying(), "Cancel shutdown preserves unrelated reminder");
        now += 29999; audio.Tick(); Check(audio.IsPlaying(), "Ordinary alert before cap"); ++now; audio.Tick(); Check(!audio.IsPlaying() && !out->playing, "Ordinary alert capped at 30 seconds");
        audio.Play(Kind::Shutdown); now += 10000; audio.Tick(); Check(!audio.IsPlaying(), "Shutdown capped at 10 seconds"); audio.Play(Kind::Shutdown); audio.StopShutdown(); Check(!audio.IsPlaying(), "Shutdown cancel/fire stops warning");
        audio.Play(Kind::Reminder, true); now += 8000; audio.Tick(); Check(!audio.IsPlaying(), "Preview capped at 8 seconds");
        audio.Play(Kind::Reminder); out->playing = false; audio.Tick(); Check(!audio.IsPlaying(), "Ended media released without looping");
        out->fail = true; audio.Play(Kind::Reminder); Check(fallback == 2 && !audio.IsPlaying() && !audio.LastError().empty(), "Decoder failure fallback"); out->fail = false;
        DeleteFileW(mp3.c_str()); audio.Play(Kind::Reminder); Check(fallback == 3 && !audio.IsPlaying() && audio.Path(Kind::Reminder) == mp3, "Missing file fallback retains choice");
        audio.Play(Kind::Countdown); audio.SetVolumePercent(0); int starts = out->starts; audio.Play(Kind::Shutdown); Check(!audio.IsPlaying() && out->starts == starts && fallback == 3, "Mute stops and suppresses sound");
        audio.SetVolumePercent(200); Check(audio.VolumePercent() == 100, "Volume upper bound"); audio.SetVolumePercent(-1); Check(audio.VolumePercent() == 0, "Volume lower bound"); Check(audio.SetPath(Kind::Countdown, L"") && audio.Path(Kind::Countdown).empty(), "Reset selection");
        { Output* safeOut = new Output(); int safeFallback = 0; fi::AlertAudio safe(root, true, std::unique_ptr<fi::AlertAudioOutput>(safeOut), [&] { ++safeFallback; }, [&] { return now; }); safe.Play(Kind::Countdown); safe.Play(Kind::Shutdown, true); Check(!safeOut->starts && !safeFallback, "Safe mode suppresses real/default/preview audio"); }
        std::cout << "PASS: " << checks << " native alert audio checks; injected playback only, no real audio.\n"; return 0;
    } catch (const std::exception& e) { std::cerr << "FAIL: " << e.what() << "\n"; return 1; }
}
